### PR TITLE
docs: add YAGNI and comment-density guidance

### LINKED ISSUE #43: docs: no guidance against over-engineering or comment bloat (58-line feature shipped as 343 lines, 40-66% comments)
## Symptom

A human reviewer, on a batch of agent-authored PRs in an infrastructure repo, said:

> wow, that was so overengineered i had a hard time getting through it

Two distinct problems sit behind that sentence: the code did more than anyone asked for, and it was buried under comments.

## Measurements

**Volume.** For the same feature (a Puppet profile that builds a Go binary and installs it), the human reviewer's version was **58 lines**; the agent-authored version was **343**. Same job. The extra 285 lines were parameters, validation and defensive handling no consumer had asked for.

**Comment density**, measured on the added lines of that batch:

| Change | Ratio |
|---|---|
| `puppet/site/profile/manifests/service/golang.pp` | 214 comment lines to 111 code lines (**66%**) |
| a one-line config change | a four-line comment (**80%**) |
| three other PRs | 44%, 40%, 37% |

**Concrete shapes of the over-engineering** in that batch:

- Class parameters with exactly one call site, or none: configurability invented for a hypothetical second consumer.
- Elaborate `Pattern[...]` type constraints validating inputs that cannot occur, alongside `assert_type` guards for the same.
- Bespoke machinery (a content-keyed build stamp with an invalidation exec) where the consuming system runs the code exactly once per image build.
- Defensive handling of states unreachable given the caller.

## The gap

Nothing in the current guidance sets an expectation about comment density, so none of that violated a rule. The closest existing line, in `coding-standards.md` under `## Documentation`, is directionally right but far too weak to prevent it:

> Inline comments only for non-obvious logic — document the *why*, not the *what*

It permits a comment wherever logic is arguably non-obvious, and says nothing about length, volume, or what to delete. `git-workflow.md`'s pre-commit review dimensions cover comment *accuracy* ("did they rot during edits?") but not comment *density* or over-engineering, so the review gate does not catch either.

On the over-engineering side the guidance is not merely absent, it arguably points the wrong way. `CLAUDE.md` §5 "Demand Elegance (Balanced)" opens with:

> For non-trivial changes: pause and ask "is there a more elegant way?"

Read cold, that is an invitation to elaborate. It plausibly contributed to this outcome. `CLAUDE.md` already has "Simplicity First" and "Minimal Impact", but neither says anything about YAGNI, speculative configurability, or premature abstraction, and neither is strong enough to outweigh a standing instruction to seek elegance.

## Proposed direction

**1. A `## Simplicity & Scope (YAGNI)` section in `coding-standards.md`:**

- **YAGNI** — build only what a current caller needs; no parameters, flags, hooks, or abstraction layers for a future that hasn't arrived. One consumer means write it for that consumer.
- **The simplest thing that works wins** — fewer lines, fewer concepts, fewer names. Sophistication is a cost, not a virtue.
- **DRY, but rule-of-three** — two similar pieces of code are usually fine; abstract at the third. Premature abstraction is over-engineering too, and is harder to undo than duplication.
- **Don't validate the impossible** — validate at trust boundaries; inside the module, trust your own callers rather than re-checking what the type system or the single call site already guarantees.
- A short **pre-PR self-check**: does every parameter have a caller? does every abstraction have more than one consumer? does every validation guard a reachable state? could a competent colleague have written this in half the lines?

**2. A `## Comments` section in `coding-standards.md`:**

- **Default to no comment** — naming and structure carry the meaning.
- **Comment only where the WHY is not deducible from the code** — a deliberate choice a reader would otherwise "fix", a constraint invisible at that spot, a workaround for an external bug, a deviation from the obvious approach.
- **Terse: 1-2 lines**, never paragraphs. If it needs more, it belongs in the PR description or commit message.
- **Rough budget: under ~15% comment-to-code on added lines** — a self-check, not a hard gate.
- **Explicitly delete**: restatements of the next line; multi-paragraph rationale essays; what-was-considered-and-rejected narrative; references to review rounds, past bugs, or "as of <date>"; JSDoc/docstrings restating types the signature already gives; long preambles on tests.
- Framing worth stating outright: **rationale belongs in the PR description; the source carries only what a future editor needs in order not to break something.**

**3. Reconciliation, so nothing contradicts:**

- Clarify `CLAUDE.md` §5 in place: elegance means **fewer moving parts**, not more sophisticated ones. A prompt to remove machinery, never to add it.
- Fold YAGNI into the existing "Simplicity First" core principle rather than bolting on a competing bullet, and add a one-line comment-density principle.
- Replace the weak `## Documentation` inline-comments line with a pointer, so there is a single source of truth.
- Add over-engineering and comment-density to `git-workflow.md`'s pre-commit review dimensions, so the review gate actually catches both.

Deliberately proportionate: two tight sections plus a short checklist, not a page. The same restraint the guidance asks for should be visible in the guidance itself.

### PR DESCRIPTION
**Pull request by cristim**
Closes #43

## Why

A reviewer on a batch of agent-authored PRs: *"wow, that was so overengineered i had a hard time getting through it."* Two problems sat behind that sentence, and no rule covered either one.

**Volume.** For the same feature (a Puppet profile that builds a Go binary and installs it), the reviewer's version was 58 lines; the agent's was 343.

**Comment bloat on top of it.** Measured on added lines: 66% comment-to-code on the worst file (214 comment lines to 111 code lines), a four-line comment on a one-line config change, and three other PRs at 44%, 40% and 37%.

A line-by-line audit of that manifest is what shaped the guidance, because the honest result was not "the extra 285 lines were fat":

- **About a third was load-bearing** and the shorter version would fail without it: a `binary_name` parameter a real caller needs, `GOPRIVATE` + submodule auth for private module fetches, `timeout => 1800` where Puppet's exec default of 300s is wrong, running the build as the checkout owner, an `ensure: absent` guard.
- **The rest was real over-engineering** with a recognisable shape: a parameter no caller overrides and one whose single caller always takes the default; ~15 lines of content-keyed build-stamp idempotence machinery for a second run that cannot happen (Packer runs Puppet once per image, then the toolchain is deleted); `Pattern[...]` constraints *plus* `assert_type` guards for the same values.

The build-stamp machinery is the instructive case: **correct, carefully reasoned, and tested, and still waste**, because the state it guarded was unreachable. That is what the section is built around. Reworked, the manifest went 343 -> 100 lines with every load-bearing behaviour intact, and comments 214 -> 20.

## What changed

**`coding-standards.md`** gains two sections:

- **`## Simplicity & Scope (YAGNI)`** — leads with the test that the audit produced: *not* "is this correct?" or "could this ever happen?" but **"does a current caller need this, and can this state be reached given the callers that exist?"** Then: build only what a current caller needs; DRY but rule-of-three, since premature abstraction is over-engineering too and harder to undo than duplication; validate at trust boundaries and don't double up; **prefer removing the hazard to defending against it** (if a value needs validating because it reaches a shell, check whether it needs to reach a shell at all). Ends with a four-question pre-PR self-check and a closing caution that cutting means removing speculation, not behaviour — "shorter" is the symptom of getting this right, not the goal.
- **`## Comments`** — default to no comment; comment only where the WHY isn't deducible from the code; 1-2 lines, never paragraphs; a ~15% comment-to-code budget on added lines as a self-check; an explicit delete-on-sight list (restatements, rationale essays, review-round references, JSDoc restating the signature). Framing: rationale belongs in the PR description, the source carries only what a future editor needs in order not to break something. One good/bad example pair.

## How the existing guidance was reconciled

This is the part worth reviewing, since the risk was bolting contradicting rules alongside what's already there.

- **`CLAUDE.md` §5 "Demand Elegance (Balanced)"** opened with *"pause and ask 'is there a more elegant way?'"*, which reads cold as an invitation to elaborate and plausibly contributed to the outcome above. Rather than adding a competing rule, §5 is clarified **in place**: the question becomes "is there a simpler way?", and elegance is defined as **fewer moving parts, not more sophisticated ones** — explicitly a prompt to remove machinery, never to add it. Section number and heading are unchanged, so the existing `§5` cross-references still resolve.
- **"Simplicity First"** absorbs YAGNI instead of gaining a neighbouring bullet saying nearly the same thing. Only one genuinely new core principle is added (comment sparingly), matching the shape of the existing entries.
- The weak `## Documentation` line (*"Inline comments only for non-obvious logic"*) becomes a pointer to the new `## Comments` section, leaving one source of truth instead of two overlapping statements.
- **`git-workflow.md`** pre-commit review dimensions gain "Over-engineering & scope" and extend "Comment accuracy" to "Comment accuracy & density" — the gate previously checked whether comments had rotted but never whether there were too many. Without this the new standards would have had no enforcement point.

## Scope

52 insertions, 6 deletions across three files. Kept deliberately tight: writing a treatise about not writing treatises would undercut the point.

## Summary by CodeRabbit

* **Documentation**
  * Added guidance on keeping changes simple, focused, and free from unnecessary abstractions.
  * Expanded recommendations for writing concise, accurate, and useful comments.
  * Added practical checks for validating scope, handling hazards, and reviewing documentation.
  * Updated the pre-commit checklist to include over-engineering, scope, and comment quality.
  * Clarified when to introduce abstractions and how to avoid speculative parameters and unnecessary machinery.

### CODE REVIEW

**Comment by cristim** [comment]
@coderabbitai review

File: coding-standards.md [hunk 1]
```diff
@@ -141,10 +156,34 @@ Ensure `.github/workflows/` (or equivalent) has at minimum:
...
+// We use a 30 second timeout here. This value was chosen after considering
+// 10s (too aggressive, caused flakes in CI) and 60s (too slow to fail).
+// 30s is a good middle ground balancing responsiveness against reliability.
+// Note: raised from 10s in review round 2.
+// See PR #412 for the full discussion.
+ctx, cancel := context.WithTimeout(ctx, 30*time.Second)
+
+// Good: one line, and only because the number is otherwise unexplainable
+// Upstream's p99 is 25s; anything tighter flakes.
+ctx, cancel := context.WithTimeout(ctx, 30*time.Second)
```
[the remark this answers is not in the document]
Comment (cristim) [reply coding-standards.md:186]: Fixed in 08e6ede: `defer cancel()` added after both `context.WithTimeout` calls. Good catch, `cancel` was unused so the snippet would not have compiled.

**Comment by cristim** [comment]
Triage of the review, thanks @coderabbitai:

**Addressed in 08e6ede**
- `defer cancel()` added to both `context.WithTimeout` examples (they would not have compiled as written).
- "Delete on sight" nitpick, external-bug half: the list said "past bugs" while the bullet above sanctions a comment for an external-bug workaround. Now "our own past bugs", and the sanctioned case explicitly says to link the upstream issue or CVE.

**Dismissed**
- "Delete on sight", JSDoc half: the entry is already scoped to "JSDoc/docstrings *restating types the signature already gives*". Semantic API docs are not covered by it and the Documentation section still requires them, so there is nothing to reconcile.
- "Align the simplicity heuristics" (qualify the added-concept rule, soften rule-of-three to a consideration, drop the line-count and consumer tests): this reverses the change rather than refining it. The PR exists because "elegance" was being read as an invitation to add machinery, so the whole point of the §5 wording is that an added concept is a cost that has to earn itself. The counterweight you are asking for is already in the section: "Cut speculation, not behaviour. A parameter a real caller sets, a timeout the platform's default gets wrong, the auth config a private dependency needs: all load-bearing, however many lines they add." The self-check items are deliberately blunt prompts, not gates, and softening every one of them into "evaluate current need, correctness, clarity and total complexity" would leave guidance that no longer decides anything. Also declining on scope grounds: this is a PR about not over-writing, and the fix would grow it substantially.

**Comment by cristim** [comment]
@coderabbitai full review

**Comment by cristim** [comment]
@coderabbitai full review

File: coding-standards.md [hunk 2]
```diff
@@ -141,10 +161,36 @@ Ensure `.github/workflows/` (or equivalent) has at minimum:
...
+
+- **Default to no comment.** Naming and structure carry the meaning; reach for a comment only where they can't.
+- **Comment only where the WHY isn't deducible from the code**: a deliberate choice a reader would otherwise "fix", a constraint that isn't visible at this spot, a workaround for an external bug (link the upstream issue or CVE), a deviation from the obvious approach.
+- **Keep it terse: 1-2 lines.** Never paragraphs, never half a page. If it needs more than that, it belongs in the PR description or the commit message.
+- **Rough budget: under ~15% comment-to-code on added lines.** A guideline for self-checking, not a hard gate; config files and public API surfaces legitimately run higher.
+- **Delete on sight**: restatements of the next line; multi-paragraph rationale essays; what-was-considered-and-rejected narrative; references to review rounds, our own past bugs, or "as of <date>"; JSDoc/docstrings restating types the signature already gives; long preambles on tests (the test name carries the meaning).
```
[the remark this answers is not in the document]
Comment (cristim) [reply coding-standards.md]: Fixed in ce5de20 by wrapping the phrase in backticks. Verified the rendered bullet really was dropping the placeholder.

File: git-workflow.md [hunk 3]
```diff
@@ -44,7 +44,8 @@ Before every commit, enter a review loop (same discipline as the plan review loo
 - **Silent failures** — swallowed errors, empty catch blocks, fallbacks that mask real problems, `nil`/zero placeholders standing in for absent data (per `CLAUDE.md` §5).
 - **Test coverage & edge cases** — are the new paths actually exercised, including boundaries, error paths, and the contract (not just the happy path)?
 - **Security** — beyond OWASP basics: trust boundaries, authz on every new path, secret handling, injection via every new input.
-- **Comment & doc accuracy** — do comments match the code, or did they rot during edits?
+- **Over-engineering & scope** — parameters with no caller, abstractions with one consumer, validation of unreachable states, machinery the current requirement doesn't need. Could a competent colleague have written this in half the lines? See `coding-standards.md` ("Simplicity & Scope (YAGNI)").
+- **Comment accuracy & density** — do comments match the code, or did they rot during edits? Is the diff over-commented (restatements of the next line, rationale essays, review-round references)? See `coding-standards.md` ("Comments").
```
[the remark this answers is not in the document]
Comment (cristim) [reply git-workflow.md:48]: Declining this one, on scope rather than on the merits.

This PR adds two bullets to the existing checklist (Over-engineering & scope, Comment accuracy & density). The "NO EXCEPTIONS" heading and the "Do NOT skip, shortcut, or batch this step" sentence are pre-existing text that the diff does not touch, so restating the skip policy here would be a drive-by edit to guidance this PR was not opened to change. It would also be a substantive policy change, not a wording fix: scoping the pre-commit review loop to "non-trivial or substantial changes" narrows when the loop is mandatory, which is the author's call and not something to fold into a docs PR about comment density.

For the record, I do not think the two documents actually conflict. `CLAUDE.md` §1b and §1c let a trivially mechanical edit skip the *worktree* and the *local Opus review loop*; the "NO EXCEPTIONS" here is about not skipping or batching passes once you are committing, which is a different gate. If the tension is worth resolving anyway it should be its own issue against the pre-existing wording, and I am happy to open one.

**Comment by cristim** [comment]
@coderabbitai full review

**Comment by cristim** [comment]
@coderabbitai full review
